mypyc/irbuild/specialize.py

```diff
@@ -138,6 +139,19 @@ def translate_globals(builder: IRBuilder, expr: CallExpr, callee: RefExpr) -> Va
     return None
 
 
+@specialize_function("builtins.abs")
+def translate_abs(builder: IRBuilder, expr: CallExpr, callee: RefExpr) -> Value | None:
+    """Specialize calls on native classes that implement __abs__."""
+    if len(expr.args) == 1 and expr.arg_kinds == [ARG_POS]:
+        arg = expr.args[0]
+        arg_typ = builder.node_type(arg)
+        if isinstance(arg_typ, RInstance) and arg_typ.class_ir.has_method("__abs__"):
+            obj = builder.accept(arg)
+            return builder.gen_method_call(obj, "__abs__", [], None, expr.line)
```

Comment:
What happens if `__abs__` has been defined incorrectly and, for example, accepts an argument? Maybe guard against this.

Reply:
Wouldn't that fail type checking as `abs()` is annotated to only accept an object that implements `__abs__` in accordance to: 

```python
@runtime_checkable
class SupportsAbs(Protocol[_T_co]):
    @abstractmethod
    def __abs__(self) -> _T_co: ...
    
abs(__x: SupportsAbs[_T]) -> _T: ...
```

https://github.com/python/typeshed/blob/2c534dc2208191dc1ae1a6f51554fa5022cbfa0a/stdlib/typing.pyi#L313-L316
https://github.com/python/typeshed/blob/2c534dc2208191dc1ae1a6f51554fa5022cbfa0a/stdlib/builtins.pyi#L1199

```python
class A:
    def __abs__(self, arg: int) -> int:
        return arg + 1


v = abs(A())
```
```
❯ mypyc t.py
t.py:6: error: Need type annotation for "v"
t.py:6: error: Argument 1 to "abs" has incompatible type "A"; expected "SupportsAbs[<nothing>]"
```
There's also a if guard that checks for _one_ positional argument before attempting specialization so even if this somehow type checks, the specialization will fail and return None. The return type shouldn't matter as `abs()` doesn't check the return type of `__abs__` at all. 

Am I missing something?

Reply:
I do see that mypyc doesn't complain if you don't ever call `abs()` on the native class with the invalid `__abs__` dunder and simply generates bad code that isn't compatible with the C-API:

```
build/__native.c:39:20: error: initialization of ‘PyObject * (*)(PyObject *)’ {aka ‘struct _object * (*)(struct _object *)’} from incompatible pointer type ‘PyObject * (*)(PyObject *, PyObject *)’ {aka ‘struct _object * (*)(struct _object *, struct _object *)’} [-Werror=incompatible-pointer-types]
   39 |     .nb_absolute = CPyDunder___abs__A,
      |                    ^~~~~~~~~~~~~~~~~~
build/__native.c:39:20: note: (near initialization for ‘A_as_number.nb_absolute’)
```

... but I don't think checking for invalid dunders (and raising an error) should happen here since it applies to all dunders, not just `__abs__`.

Reply:
Yeah, this is something that should be checked elsewhere. I'll create an issue about this -- no need to fix it here.